dnn/utils/yolo_encoder.py

```diff
@@ -0,0 +1,132 @@
+            tf.cast(y_true_n6[...,5] % 3, 
+            dtype=tf.int32)], axis=1)
+        
+        #update should contain object info, size (n, 6)
+        #(x1, y1, x1, y1, obj, class)
+        #!TODO precalculate xy center and hw instead of doing it in loss function
+        updates_n6 = tf.stack(
+            [box_yxyx_n4[...,1], 
+            box_yxyx_n4[...,0], 
+            box_yxyx_n4[...,3],
+            box_yxyx_n4[...,2],
+            tf.ones((n,)), 
+            y_true_n6[...,4]], axis=1)
+
+        out = tf.tensor_scatter_nd_update(y_true_out_gg36, indexes_n3, updates_n6)
+        return out
+
+    def transform_label(self, y_train_n5):
+        '''
+        Read raw y_train and return yolo_out of different dimensions.
+
+        args:
+            y_train_n5: label, tensor of shape (n, (x,y,x,y,class)), should be in range(0, 1), 
+                    bounding box index must be normalized with image size
+        '''
+        y_outs = []
+        grid_size = self.img_size // 32
+
+        # calculate anchor index for true boxes
+        anchors_n2 = tf.cast(self.anchors_n2, tf.float32)
+        anchor_area = anchors_n2[..., 0] * anchors_n2[..., 1]
+
+        # y_true is yxyx after imgaug
+        # How to calculate hw depends on tf_record format
+        box_hw = y_train_n5[..., 2:4] - y_train_n5[..., 0:2]
+        box_hw = tf.tile(tf.expand_dims(box_hw, -2),
+                        (1, tf.shape(anchors_n2)[0], 1))
+        box_area = box_hw[..., 0] * box_hw[..., 1]
+
+        # Calculate intersection between bbox and anchors
+        intersection = tf.minimum(box_hw[..., 1], anchors_n2[..., 0]) * \
+            tf.minimum(box_hw[..., 0], anchors_n2[..., 1])
+
+        # Calculate iou, which is intersection/total area
+        iou = intersection / (box_area + anchor_area - intersection)
+        
+        # Find the best anchor
+        anchor_idx = tf.cast(tf.argmax(iou, axis=-1), tf.float32)
+        anchor_idx = tf.expand_dims(anchor_idx, axis=-1)
+
+        # n*6
+        y_train_n6 = tf.concat([y_train_n5, anchor_idx], axis=-1)
+
+        for masks in self.anchor_masks_n3:
+            y_outs.append(self.transform_label_for_output(
+                y_train_n6, grid_size, masks))
+            grid_size *= 2
+        
+        #y_out[0] = 13*13*3*6, y_out[1] = 26*26*3*6, y_out[2] = 52*52*3*6
+        # grid*grid*anchor*(x,y,w,h,confidence,class(2))
+        return tuple(y_outs)
+
+    # Wrapper for lambda function
+    def transform_images_train(self, x_train):
+        return x_train
+
+    def __call__(self, tf_record_dataset):
+        '''
+        Takes a tf_record_dataset and return the encoded dataset
+        Args:
+            tf_record_dataset: a processed tf_record, i.e., the input
+                                shape should match and y should be in 0~1
+        '''
+        return tf_record_dataset.map(lambda x, y: (
+            self.transform_images_train(x),
+            self.transform_label(y)
+        ))
```

Comment:
```suggestion
        return tf_record_dataset.map(lambda x, y: (
            self.transform_images_train(x),
            self.transform_label(y)
        ), num_parallel_calls=tf.data.AUTOTUNE)
```
Encoder is fairly expensive, try running it in parallel

Reply:
doesn't seem to be very helpful